maintenance/migratestatus/migrator/migrator.go

```diff
@@ -0,0 +1,281 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migrator
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/google/go-github/github"
+	"k8s.io/test-infra/maintenance/githubutil"
+)
+
+var (
+	stateAny     = "ANY_STATE"
+	stateSuccess = "success"
+
+	deprecatedDesc = "This status context is deprecated."
+)
+
+// contextCondition is a struct that describes a condition about the state or existance of a context.
+type contextCondition struct {
+	// context is the status context that this condition applies to.
+	context string
+	// state is the status state that the condition accepts, "ANY_STATE", or nil to accept only if
+	// the status is not present.
+	state *string
+}
+
+// Mode is a struct that describes the behavior of a status migration. The behavior is described as
+// a list of conditions and a function that determines the actions to be taken when the conditions
+// are met.
+type Mode struct {
+	conditions []*contextCondition
+	// actions returns the status updates to make based on the current statuses and the sha.
+	// When actions is called, the Mode may assume that it's conditions are met.
+	actions func(statuses []github.RepoStatus, sha string) []*github.RepoStatus
+}
+
+// ReplaceMode creates a mode that both duplicates and deprecates.
+// The mode creates a new context on every PR with the old context but not the new one, setting the
+// state of the new context to that of the old context before deprecating the old context. The
+// description and target URL of the new context are determined by the ContextDefinition.
+func ReplaceMode(origContext, newContext string) Mode {
+	dup := duplicateAction(origContext, newContext)
+	dep := deprecateAction(origContext)
+
+	return Mode{
+		conditions: []*contextCondition{
+			&contextCondition{context: origContext, state: &stateAny},
+			&contextCondition{context: newContext, state: nil},
+		},
+		actions: func(statuses []github.RepoStatus, sha string) []*github.RepoStatus {
+			return append(dup(statuses, sha), dep(statuses, sha)...)
+		},
+	}
+}
+
+// DuplicateMode makes a mode that creates a new context in every PR with the old context and without the new one.
+// The state, description and target URL of the new context are made the same as those of the old context.
+func DuplicateMode(origContext, newContext string) Mode {
+	return Mode{
+		conditions: []*contextCondition{
+			&contextCondition{context: origContext, state: &stateAny},
+			&contextCondition{context: newContext, state: nil},
+		},
+		actions: duplicateAction(origContext, newContext),
+	}
+}
+
+// DeprecateMode creates a mode that deprecates old contexts on all PRs with both the new and old context.
```

Comment:
This function creates a mode that deprecates the old context on all PRs that have both the old context and the new context. Here deprecate means set state to success and set description to a deprecation notice.  I updated the comment to try to make this clearer.